### ISSUE TITLE
An issue in CpuNeighborList.cpp

### ISSUE THREAD

**Issue by lhycms** [openmm/openmm#4141]
I need to do something based on neighbor list. So i try to understand the implementation of CpuNeighborList in https://github.com/openmm/openmm/blob/master/platforms/cpu/src/CpuNeighborList.cpp.  I can't understand below part. 

I think the fourth to last line should be `yperiodic -= periodicBoxVectors[1][1]*scale1;`, not `yperiodic -= periodicBoxVectors[1][0]*scale1;`.

Key part of CpuNeighborList.cpp is shown:
```c++
/**
     * Get the voxel index containing a particular location.
     */
    VoxelIndex getVoxelIndex(const float* location) const {
        float yperiodic, zperiodic;
        if (!usePeriodic) {
            yperiodic = location[1]-miny;
            zperiodic = location[2]-minz;
        }
        else {
            float scale2 = floorf(location[2]*recipBoxSize[2]);
            yperiodic = location[1]-periodicBoxVectors[2][1]*scale2;
            zperiodic = location[2]-periodicBoxVectors[2][2]*scale2;
            float scale1 = floorf(yperiodic*recipBoxSize[1]);
            yperiodic -= periodicBoxVectors[1][0]*scale1;
        }
        int y = max(0, min(ny-1, int(floorf(yperiodic / voxelSizeY))));
        int z = max(0, min(nz-1, int(floorf(zperiodic / voxelSizeZ))));
        
        return VoxelIndex(y, z);
```

**Comment by RaulPPelaez** [comment]
To give you a little more context about the box vectors, in case it helps:
 
The box vectors `a`, `b`, and `c` must satisfy certain requirements:
        `a[1] = a[2] = b[2] = 0`
        `a[0] >= 2*cutoff, b[1] >= 2*cutoff, c[2] >= 2*cutoff`
        `a[0] >= 2*b[0]`
        `a[0] >= 2*c[0]`
        `b[1] >= 2*c[1]`

These requirements correspond to a particular rotation of the system and reduced form of the vectors, as well as the requirement that the cutoff be no larger than half the box width.

**Comment by lhycms** [comment]
> I need to do something based on neighbor list. So i try to understand the implementation of CpuNeighborList in https://github.com/openmm/openmm/blob/master/platforms/cpu/src/CpuNeighborList.cpp. I can't understand below part.
> 
> I think the fourth to last line should be `yperiodic -= periodicBoxVectors[1][1]*scale1;`, not `yperiodic -= periodicBoxVectors[1][0]*scale1;`.
> 
> Key part of CpuNeighborList.cpp is shown:
> 
> ```c++
> /**
>      * Get the voxel index containing a particular location.
>      */
>     VoxelIndex getVoxelIndex(const float* location) const {
>         float yperiodic, zperiodic;
>         if (!usePeriodic) {
>             yperiodic = location[1]-miny;
>             zperiodic = location[2]-minz;
>         }
>         else {
>             float scale2 = floorf(location[2]*recipBoxSize[2]);
>             yperiodic = location[1]-periodicBoxVectors[2][1]*scale2;
>             zperiodic = location[2]-periodicBoxVectors[2][2]*scale2;
>             float scale1 = floorf(yperiodic*recipBoxSize[1]);
>             yperiodic -= periodicBoxVectors[1][0]*scale1;
>         }
>         int y = max(0, min(ny-1, int(floorf(yperiodic / voxelSizeY))));
>         int z = max(0, min(nz-1, int(floorf(zperiodic / voxelSizeZ))));
>         
>         return VoxelIndex(y, z);
> ```



> To give you a little more context about the box vectors, in case it helps:
> 
> The box vectors `a`, `b`, and `c` must satisfy certain requirements: `a[1] = a[2] = b[2] = 0` `a[0] >= 2*cutoff, b[1] >= 2*cutoff, c[2] >= 2*cutoff` `a[0] >= 2*b[0]` `a[0] >= 2*c[0]` `b[1] >= 2*c[1]`
> 
> These requirements correspond to a particular rotation of the system and reduced form of the vectors, as well as the requirement that the cutoff be no larger than half the box width.

Thanks for your help! These information really help a lot! 
The restriction `cutoff ≤ boxSize/2` is really terrible.  I think there may be some ways to remove this restriction.
But i can't understand why we must guaratee `a[0] >= 2*b[0]`, `a[0] >= 2*c[0]` and `b[1] >= 2*c[1]`.

**Comment by RaulPPelaez** [comment]
Read the start here:
https://docs.lammps.org/Howto_triclinic.html

Here you have a much more in depth discussion, see about PCT1:
/https://doi.org/10.1002/(SICI)1096-987X(19971130)18:15<1930::AID-JCC8>3.0.CO;2-P

**Comment by RaulPPelaez** [comment]
See, given the a,b,c box vectors, the transformation matrix between the cartesian coordinates and the triclinic coordinates is:
![image](https://github.com/openmm/openmm/assets/13015792/ede396d8-9cdd-4d62-add8-07f74854846c)

You need the inverse of this, which is straight-forward to lie down given the zeroes, and multiply by the atom coordinates:
![image](https://github.com/openmm/openmm/assets/13015792/630c9ac0-ad98-4254-b78a-115bc1fb85b4)

There you can see the "xy" factor, ```periodicBoxVector[0][1]``` in the code.
Warning: Completely, quick and dirty, unchecked math

**Comment by lhycms** [comment]
Thanks for you explaination!
In addition to above main issue about CpuNeighborList, I also have one more question. How Python code call C++code in OpenMM, it seems that OpenMM doesn't use pybind11 wrapper.

**Comment by peastman** [comment]
> The restriction cutoff ≤ boxSize/2 is really terrible.

That requirement is critical.  Without it, a point could be within the cutoff distance of multiple periodic copies of a single atom.  That would make computing nonbonded interactions much more difficult and much less efficient.

>  But i can't understand why we must guaratee `a[0] >= 2*b[0], a[0] >= 2*c[0] and b[1] >= 2*c[1]`.

It makes the math easier in a lot of places.  That just corresponds to a particular choice of reduced box vectors.  Given arbitrary a, b, and c that don't satisfy those requirements, the following code puts them into an equivalent form (representing the same periodic system) that does.

https://github.com/openmm/openmm/blob/806e79c51f05a2f827b7d571eb6a4b4fc721fe8f/wrappers/python/openmm/app/internal/unitcell.py#L95-L97

> it seems that OpenMM doesn't use pybind11 wrapper

The Python wrapper is generated with SWIG.

**Comment by lhycms** [comment]
Thank you very much for your help. But the initial problem has not been resolved yet.

> I think the fourth to last line should be `yperiodic -= periodicBoxVectors[1][1]*scale1;`, not `yperiodic -= periodicBoxVectors[1][0]*scale1;`.

Could you help check above expression？

**Comment by peastman** [comment]
I believe you are correct about that.  Good catch!  A PR would be welcome!

**Comment by peastman** [comment]
Fixed by #4144.
